modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessPlugin.java

```diff
@@ -115,7 +116,8 @@ public ScriptEngine getScriptEngine(Settings settings, Collection<ScriptContext<
                                                ResourceWatcherService resourceWatcherService, ScriptService scriptService,
                                                NamedXContentRegistry xContentRegistry, Environment environment,
                                                NodeEnvironment nodeEnvironment, NamedWriteableRegistry namedWriteableRegistry,
-                                               IndexNameExpressionResolver expressionResolver) {
+                                               IndexNameExpressionResolver expressionResolver,
+                                               Supplier<RepositoriesService> repositoriesServiceSupplier) {
```

Comment:
component creation (ie services created by plugins) should happen almost last within node initialization. I don't think we should be passing in a Supplier. Instead, we should move the call to createComponents down later, so that we have the RepositoriesService.

Reply:
I tried this (see b2b5333) but e.g. `XPackPlugin` requires us to call `createComponents` before initialising `ActionModule`:

```
»  org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: SSL Service is not constructed yet |  
-- | --
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:174) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:161) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:127) ~[elasticsearch-cli-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.cli.Command.main(Command.java:90) ~[elasticsearch-cli-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:126) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:92) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  Caused by: java.lang.IllegalStateException: SSL Service is not constructed yet |  
  | »  	at org.elasticsearch.xpack.core.XPackPlugin.getSharedSslService(XPackPlugin.java:162) ~[?:?] |  
  | »  	at org.elasticsearch.xpack.security.Security.getSslService(Security.java:339) ~[?:?] |  
  | »  	at org.elasticsearch.xpack.security.Security.getRestHandlerWrapper(Security.java:980) ~[?:?] |  
  | »  	at org.elasticsearch.action.ActionModule.<init>(ActionModule.java:455) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.node.Node.<init>(Node.java:468) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.node.Node.<init>(Node.java:256) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:225) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:225) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:387) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:170) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT] |  
  | »  	... 6 more
```

We need the `ActionModule` to create the `TransportService` and we need a `TransportService` to create the `RepositoriesService`, so we seem to be stuck.

Reply:
Thanks for trying, @DaveCTurner. The action module part is avoidable by tweaking when the security rest wrapper invokes `getSslService()`, but the aggressive creation of transport doesn't seem to be resolvable, at least for now. This is all a result of the odd sharing of the ssl service across xpack plugins. I'll take this as a TODO to investigate how we can clean this up with other plugin/node initialization cleanup work.